specs/_features/whisk/beacon-chain.md

```diff
@@ -0,0 +1,490 @@
+```
+
+Removed `assert block.proposer_index == get_beacon_proposer_index(state)` check in Whisk.
+
+```python
+def process_block_header(state: BeaconState, block: BeaconBlock) -> None:
+    # Verify that the slots match
+    assert block.slot == state.slot
+    # Verify that the block is newer than latest block header
+    assert block.slot > state.latest_block_header.slot
+
+    # # Verify that proposer index is the correct index
+    # assert block.proposer_index == get_beacon_proposer_index(state)
+
+    # Verify that the parent matches
+    assert block.parent_root == hash_tree_root(state.latest_block_header)
+    # Cache current block as the new latest block
+    state.latest_block_header = BeaconBlockHeader(
+        slot=block.slot,
+        proposer_index=block.proposer_index,
+        parent_root=block.parent_root,
+        state_root=Bytes32(),  # Overwritten in the next process_slot call
+        body_root=hash_tree_root(block.body),
+    )
+
+    # Verify proposer is not slashed
+    proposer = state.validators[block.proposer_index]
+    assert not proposer.slashed
+    process_whisk_opening_proof(state, block)   # [New in Whisk]
+```
+
+### Whisk
+
+#### `BeaconBlockBody`
+
+```python
+class BeaconBlockBody(capella.BeaconBlockBody):
+    randao_reveal: BLSSignature
+    eth1_data: Eth1Data  # Eth1 data vote
+    graffiti: Bytes32  # Arbitrary data
+    # Operations
+    proposer_slashings: List[ProposerSlashing, MAX_PROPOSER_SLASHINGS]
+    attester_slashings: List[AttesterSlashing, MAX_ATTESTER_SLASHINGS]
+    attestations: List[Attestation, MAX_ATTESTATIONS]
+    deposits: List[Deposit, MAX_DEPOSITS]
+    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
+    sync_aggregate: SyncAggregate
+    # Execution
+    execution_payload: ExecutionPayload
+    # Capella operations
+    bls_to_execution_changes: List[SignedBLSToExecutionChange, MAX_BLS_TO_EXECUTION_CHANGES]
+    whisk_post_shuffle_trackers: Vector[WhiskTracker, WHISK_VALIDATORS_PER_SHUFFLE]  # [New in Whisk]
+    whisk_shuffle_proof: WhiskShuffleProof  # [New in Whisk]
+    whisk_shuffle_proof_M_commitment: BLSG1Point  # [New in Whisk]
+    whisk_registration_proof: WhiskTrackerProof  # [New in Whisk]
+    whisk_tracker: WhiskTracker  # [New in Whisk]
+    whisk_k_commitment: BLSG1Point  # [New in Whisk]
+```
+
+```python
+def get_shuffle_indices(randao_reveal: BLSSignature) -> Sequence[uint64]:
+    """
+    Given a `randao_reveal` return the list of indices that got shuffled from the entire candidate set
+    """
+    indices = []
+    for i in WHISK_VALIDATORS_PER_SHUFFLE:
+        # XXX ensure we are not suffering from modulo bias
+        shuffle_index = uint256(hash(randao_reveal + uint_to_bytes(i))) % WHISK_CANDIDATE_TRACKERS_COUNT
+        indices.append(shuffle_index)
+
+    return indices
+```
+
+```python
+def whisk_process_shuffled_trackers(state: BeaconState, body: BeaconBlockBody) -> None:
```

Comment:
```suggestion
def process_shuffled_trackers(state: BeaconState, body: BeaconBlockBody) -> None:
```

Reply:
Almost every new function and property in this PR is prefix by `whisk_`, should not only this function but everything else drop the prefix?

Reply:
Renamed process_shuffled_trackers only